File: src/services/audit/detectors/ad/attack-paths/high/nested-admin.ts
/**
 * Nested Admin Path Detector
 *
 * Detects excessive group nesting leading to admin access.
 * Groups nested more than 3 levels deep reaching admin groups.
 */

import { ADUser, ADGroup } from '../../../../../../types/ad.types';
import { Finding } from '../../../../../../types/finding.types';
import { toAffectedUserEntities } from '../../../../../../utils/entity-converter';
import { detectNestingDepth, PRIVILEGED_GROUPS } from '../../../../../../utils/graph.util';

/**
 * Detect excessive group nesting leading to admin access
 * Groups nested more than 3 levels deep reaching admin groups
 */
export function detectPathNestedAdmin(
  users: ADUser[],
  groups: ADGroup[],
  includeDetails: boolean
): Finding {
  const EXCESSIVE_DEPTH = 3;

  // Find privileged groups
  const privilegedGroupDns = groups
    .filter((g) =>
      PRIVILEGED_GROUPS.some(
        (pg) =>
          g.sAMAccountName?.toLowerCase().includes(pg.toLowerCase()) ||
          g.dn.toLowerCase().includes(`cn=${pg.toLowerCase()}`)
      )
    )
    .map((g) => g.dn);

  // Find users in deeply nested paths to admin
  const affected: ADUser[] = [];
  const deepNestingDetails: { user: string; depth: number; group: string }[] = [];

  for (const user of users) {
    if (!user.memberOf || user.memberOf.length === 0) continue;

    for (const groupDn of user.memberOf) {
      const depth = detectNestingDepth(groupDn, groups);

      // Check if this chain reaches a privileged group
      if (depth > EXCESSIVE_DEPTH) {
        // Check if any parent in chain is privileged
        const group = groups.find((g) => g.dn.toLowerCase() === groupDn.toLowerCase());
        if (group?.memberOf?.some((parent) => privilegedGroupDns.includes(parent))) {
          affected.push(user);
          deepNestingDetails.push({
            user: user.sAMAccountName,
            depth,
            group: groupDn,
          });
          break;
        }
      }
    }
  }

  return {
    type: 'PATH_NESTED_ADMIN',
    severity: 'high',
    category: 'attack-paths',
    title: 'Excessive Group Nesting to Admin',
    description: `Users reach admin groups through excessive nesting (>${EXCESSIVE_DEPTH} levels). Makes privilege review difficult and may hide admin access.`,
    count: affected.length,
    affectedEntities: includeDetails ? toAffectedUserEntities(affected) : undefined,
    details:
      affected.length > 0
        ? {
            threshold: EXCESSIVE_DEPTH,
            deepestNesting: Math.max(...deepNestingDetails.map((d) => d.depth), 0),
            mitigation: 'Flatten group structure, limit nesting to 2-3 levels',
          }
        : undefined,
  };
}

File: internal/audit/detectors/ad/attack-paths/high/nested-admin.go
package high

import (
	"context"
	"strings"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/internal/audit/helpers"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// NestedAdminDetector detects excessive group nesting leading to admin access
type NestedAdminDetector struct {
	audit.BaseDetector
}

// NewNestedAdminDetector creates a new detector
func NewNestedAdminDetector() *NestedAdminDetector {
	return &NestedAdminDetector{
		BaseDetector: audit.NewBaseDetector("PATH_NESTED_ADMIN", audit.CategoryAttackPaths),
	}
}

const excessiveDepth = 3

var privilegedGroups = []string{
	"domain admins",
	"enterprise admins",
	"administrators",
	"schema admins",
	"account operators",
	"backup operators",
	"server operators",
}

// Detect executes the detection
func (d *NestedAdminDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	// Build group membership map
	groupMemberOf := make(map[string][]string)
	for _, g := range data.Groups {
		groupMemberOf[strings.ToLower(g.DN)] = g.MemberOf
	}

	// Check for privileged groups
	privilegedGroupDNs := make(map[string]bool)
	for _, g := range data.Groups {
		for _, pg := range privilegedGroups {
			if strings.Contains(strings.ToLower(g.SAMAccountName), pg) {
				privilegedGroupDNs[strings.ToLower(g.DN)] = true
				break
			}
		}
	}

	var affected []types.User
	deepestNesting := 0

	for _, u := range data.Users {
		if len(u.MemberOf) == 0 {
			continue
		}

		// Check depth of each group membership
		for _, groupDN := range u.MemberOf {
			depth := calculateNestingDepth(strings.ToLower(groupDN), groupMemberOf, make(map[string]bool))

			if depth > excessiveDepth {
				// Check if path reaches privileged group
				if reachesPrivileged(strings.ToLower(groupDN), groupMemberOf, privilegedGroupDNs, make(map[string]bool)) {
					affected = append(affected, u)
					if depth > deepestNesting {
						deepestNesting = depth
					}
					break
				}
			}
		}
	}

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityHigh,
		Category:    string(d.Category()),
		Title:       "Excessive Group Nesting to Admin",
		Description: "Users reach admin groups through excessive nesting (>3 levels). Makes privilege review difficult and may hide admin access.",
		Count:       len(affected),
	}

	if data.IncludeDetails && len(affected) > 0 {
		finding.AffectedEntities = helpers.ToAffectedUserEntities(affected)
		finding.Details = map[string]interface{}{
			"threshold":      excessiveDepth,
			"deepestNesting": deepestNesting,
			"mitigation":     "Flatten group structure, limit nesting to 2-3 levels",
		}
	}

	return []types.Finding{finding}
}

func calculateNestingDepth(groupDN string, groupMemberOf map[string][]string, visited map[string]bool) int {
	if visited[groupDN] {
		return 0
	}
	visited[groupDN] = true

	parents := groupMemberOf[groupDN]
	if len(parents) == 0 {
		return 1
	}

	maxDepth := 0
	for _, parent := range parents {
		depth := calculateNestingDepth(strings.ToLower(parent), groupMemberOf, visited)
		if depth > maxDepth {
			maxDepth = depth
		}
	}

	return maxDepth + 1
}

func reachesPrivileged(groupDN string, groupMemberOf map[string][]string, privilegedDNs map[string]bool, visited map[string]bool) bool {
	if visited[groupDN] {
		return false
	}
	visited[groupDN] = true

	if privilegedDNs[groupDN] {
		return true
	}

	for _, parent := range groupMemberOf[groupDN] {
		if reachesPrivileged(strings.ToLower(parent), groupMemberOf, privilegedDNs, visited) {
			return true
		}
	}

	return false
}

func init() {
	audit.MustRegister(NewNestedAdminDetector())
}
